### DISCUSSION TITLE
How do i connect to the remote node using Golang ?

### DISCUSSION THREAD

**Discussion by khanakia** [ergo-services/ergo#101] [Q&A]
Hi,

How do i connect to the remote node using Golang?  I do not see any examples of that.

For e.g. i can do in elixir like this
```
iex --sname bar
GenServer.call({B, :'foo@Amans-MacBook-Pro'}, :get_balance)
```

So I want to do the same from golang to connect to the remote node and run a function

**Comment by halturin** [comment]
Hi, thanks for the question.

You may want to try this example https://github.com/ergo-services/ergo/tree/master/examples/genserver

gen.ServerProcess has methods Cast and Call

**Reply by khanakia** [reply]
Yes, i checked that but that requires me to create a node. But on the client-side, why do I need to create a node.  
Basically, I have 2 going apps. So I want one to be a SERVER and the other to be a CLIENT so I can connect by node name from client to server.

As you can see this repo for reference they have a very simple example https://github.com/grpc/grpc-go/tree/master/examples/helloworld

Can we have something like this so it will help the users to understand easily how to create separate server and client app ?

**Comment by halturin** [comment]
in Erlang/Elixir and in Ergo as well, making any requests (sending a message, to be specific) can be done by the process only, which can exist only within a node. There is no other way (like "client-library").

**Reply by khanakia** [reply]
Yes, But how do i connect to a remote Node using another going app (CLIENT) and send request ?

PS: I can do that in elixir very easily.

Elixir example
```
## Machine 1 - SERVER
iex --name machine1@192.168.7.2 --cookie somecookie

## Machine 2 - CLIENT
iex --name machine2@192.168.7.3 --cookie somecookie
Node.connect :"machine1@192.168.7.2"
Node.list

```

**Reply by halturin** [reply]
you don't need to connect. there is a network transparent interop. Just start a node and send a message by the gen.ProcessID{Name:ProcessName, Node: NodeName} and it will make connection automatically.

if you need exactly to make a connection, use the Connect method of the node.Node interface. But I'm not sure why it could be needed.

**Reply by khanakia** [reply]
So i am using genserver as SERVER https://github.com/ergo-services/ergo/tree/master/examples/genserver

and then I created a CLIENT app here https://github.com/khanakia/ergo_examples/blob/main/client_server/client/main.go

```
func main() {
	node, _ := ergo.StartNode("node@localhost", "cookies", node.Options{})
	process, _ := node.Spawn("gs1", gen.ProcessOptions{}, &simple{})

	to := gen.ProcessID{Name: "example", Node: "demo@127.0.0.1"}
	err := process.Send(to, "hello")
	fmt.Println(err)
	// out: no route to node
}
```

But it returns an error `no route to node`

**Reply by halturin** [reply]
you used the wrong cookie value. Example uses "123" as a cookie
https://github.com/ergo-services/ergo/blob/master/examples/genserver/server.go#L66

your client has also wrong process ID value 
https://github.com/khanakia/ergo_examples/blob/main/client_server/client/main.go#L37

`to := gen.ProcessID{Name: "demo@127.0.0.1", Node: "example"}`

must be 

`to := gen.ProcessID{Node: "demo@127.0.0.1", Name: "example"}`

**Reply by khanakia** [reply]
Thanks :) It worked.

Last thing how do I send a CALL method and get the Response back because `process.Send` does not returns any response

**Reply by halturin** [reply]
The `Cast` and `Call` are `gen.ServerProcess` methods and are available within an actor only (to be specific - within the actor's callbacks). To operate with an actor from outside, you should use `gen.Process` methods: `Send` method for async messaging (will be invoked HandleInfo callback) or `Direct` to make sync calls to the actor (will be invoked HandleDirect).

* `gen.ServerProcess` is available as a first argument in the callbacks like Init, HandleCast, HandleCall etc. (see `gen.ServerBehavior` interface)
* `gen.Process` is an interface returned by Spawn method.

**Reply by khanakia** [reply]
Ok
